archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,252 @@
+lemma exists_numbers_in_interval (n : ℕ) (hn : n ≥ 107): ∃ (l : ℕ),
+  (n + 4 * l ≤ 2 * l * l ∧ 2 * l * l + 4 * l ≤ 2 * n) :=
+begin
+  suffices : ∃ (l : ℕ), 2 * (l : ℝ) ≥ 2 + sqrt (4 + 2 * n) ∧ (l : ℝ) ≤ sqrt (1 + n) - 1,
+  { cases this with l t,
+    use l,
+    refine ⟨lower_bound n l t.1, upper_bound n l t.2⟩ },
+  lift floor (sqrt (1 + n) - 1) to ℕ with x,
+  have hx : (x : ℝ) = ⌊sqrt (1 + ↑n) - 1⌋,
+    { norm_cast,
+      convert h,
+      push_cast,
+      norm_num },
+  refine ⟨x, _, _⟩,
+  { suffices : 2 * (sqrt (1 + n) - 2) ≥ 2 + sqrt (4 + 2 * n),
+    { apply le_trans this _,
+      simp only [mul_le_mul_left, zero_lt_bit0, zero_lt_one],
+      rw hx,
+      suffices : sqrt (1 + n) - 1 ≤ ↑⌊sqrt (1 + n) - 1⌋ + 1,
+      { linarith },
+      have t :  (⌈sqrt (1 + n) - 1⌉:ℝ) ≤ ⌊sqrt (1 + n) - 1⌋ + 1,
+        norm_cast,
+        exact ceil_le_floor_add_one _,
+      apply le_trans _ t,
+      exact le_ceil (sqrt (1 + ↑n) - 1) },
+      suffices : sqrt (4 + 2 * n) ≤ 2 * (sqrt (1 + n) - 3),
+      { linarith },
+    exact radical_inequality hn },
+  { rw hx,
+    exact floor_le _ },
+  { rw [floor_nonneg, le_sub, sub_zero, le_sqrt],
+    all_goals { norm_cast, linarith }},
+
+end
+
+lemma exists_triplet_summing_to_squares (n : ℕ) (hn : n ≥ 100): (∃ (a b c : ℕ),
+  n ≤ a ∧ a < b ∧ b < c ∧ c ≤ 2 * n ∧  (∃ (k : ℕ), a + b = k * k) ∧
+  (∃ (l : ℕ), c + a = l * l) ∧ (∃ (m : ℕ), b + c = m * m)) :=
+begin
+  --If n ≥ 107, we do not explicitly construct the triplet but use an existence
+  --argument from lemma above.
+  by_cases p : n ≥ 107,
+  { have h := exists_numbers_in_interval n p,
+    cases h with l hl,
+    use [2 * l * l - 4 * l, 2 * l * l + 1, 2 * l * l + 4 * l],
+    by_cases p : l > 1,
+    refine ⟨_,⟨_,⟨_,⟨_,⟨_, ⟨_, _⟩⟩⟩⟩⟩⟩,
+    { exact nat.le_sub_right_of_add_le hl.1 },
+    { exact (2 * l * l).sub_lt_succ (4 * l) },
+    { linarith },
+    { exact hl.2 },
+    { use (2 * l - 1),
+      have h₁ : 1 ≤ 2 * l, { linarith },
+      have h₂ : 4 * l ≤ 2 * l * l, { linarith },
+      zify [h₁, h₂],
+      ring_exp_eq },
+    { use (2 * l),
+      have h₁ : 4 * l ≤ 2 * l * l, { linarith },
+      zify [h₁],
+      ring_exp_eq },
+    { use(2 * l + 1),
+      ring_exp_eq },
+      --Otherwise, if 100 ≤ n < 107, then it suffices to consider explicit
+      --construction of a triplet {a, b, c}, which is constructed by setting l=9
+      --in the argument at the start of the file.
+    { exfalso,
+      simp only [not_lt] at p,
+      interval_cases l,
+      all_goals {linarith }}},
+  { refine ⟨126, 163, 198, _, _, _, _, ⟨⟨17, _⟩,⟨⟨18, _⟩,⟨19, _⟩⟩⟩⟩,
+    all_goals { linarith }},
+end
+
+lemma mem_compl_iff_nmem_subset {A : finset ℕ} {a b c n : ℕ}
+ (t : A ⊆ finset.Ico n (2 * n + 2)) (h₁ : n ≤ a)
+ (h₂ : a < b) (h₃ : b < c) (h₄ : c ≤ 2 * n) :
+ (a ∉ A ↔ a ∈ finset.Ico n (2 * n + 2) \ A) ∧
+ (b ∉ A ↔ b ∈ finset.Ico n (2 * n + 2) \ A) ∧
+ (c ∉ A ↔ c ∈ finset.Ico n (2 * n + 2) \ A) :=
```

Comment:
First, please run #lint at the bottom of this file, and make it happy.

Second, this lemma is terribly redundant. At most you want:
```
lemma foo {A : finset ℕ} {a n : ℕ} (h₁ : n ≤ a) (h₂ : a ≤ 2 * n) : 
  (a ∉ A ↔ a ∈ finset.Ico n (2 * n + 2) \ A) :=
begin
  refine ⟨_,_⟩,
  { intro ha,
    simp only [finset.Ico.mem, finset.mem_sdiff],
    refine ⟨⟨_,_⟩,ha⟩,
    all_goals { linarith }},
  { intro ha,
    simp only [finset.mem_sdiff, finset.mem_range] at ha,
    exact ha.2 },
end
```
I'm pretty confused about the `2 * n + 2` here. Why isn't it `2 * n + 1`, corresponding to `[n,2n]`?

Reply:
And even this lemma is probably better stated replacing `(h₁ : n ≤ a) (h₂ : a ≤ 2 * n)` with `(h : a ∈ finset.Ico n (2 * n + 1)`: at this point you might find it in the library, and if it's not there, put it there.

Reply:
> I'm pretty confused about the `2 * n + 2` here. Why isn't it `2 * n + 1`, corresponding to `[n,2n]`?
Oops, sorry for this silly mistake, will change it to `2 * n + 1` in the next commit.

Reply:
> And even this lemma is probably better stated replacing `(h₁ : n ≤ a) (h₂ : a ≤ 2 * n)` with `(h : a ∈ finset.Ico n (2 * n + 1)`: at this point you might find it in the library, and if it's not there, put it there.

Haven't found it in the library, I put my version of this lemma in the new commit. Now the question is, how should one name it?

Reply:
Okay, I think @YaelDillies's suggestion below, to simply not use `finset.Ico` at all in the file, is probably a good one, so this discussion will become moot.